### PR TITLE
feat:Added the db config, tenant api

### PR DESCRIPTION
**Pull request by vishwab1**
Database connection added
Tenant Api added
Response Structure added

### CODE REVIEW

**Comment by vishwab1** [comment]
@HarishGangula  Please verify the PR contain the Tenant API

File: package.json [hunk 1]
```diff
@@ -25,12 +25,14 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "@types/multer": "^1.4.12",
```
Comment (HarishGangula) [inline package.json:28]: Remove this dependecy

File: package.json [hunk 2]
```diff
@@ -25,12 +25,14 @@
...
     "@types/pg": "^8.11.6",
```
Comment (HarishGangula) [inline package.json:29]: This should go to dev dependecies

File: src/config/databaseConfig.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import 'reflect-metadata';
```
Comment (HarishGangula) [inline src/config/databaseConfig.ts:1]: Why this package used ?

File: package.json [hunk 4]
```diff
@@ -25,12 +25,14 @@
...
     "ajv": "^8.17.1",
     "cors": "^2.8.5",
     "express": "^5.0.0-beta.3",
     "http-status": "^1.7.4",
     "lodash": "^4.17.21",
+    "multer": "^1.4.5-lts.1",
     "reflect-metadata": "^0.2.2",
```
Comment (HarishGangula) [inline package.json:36]: Remove this dependency

File: src/app.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+const app: Application = express();
+
+const initializeServer = async (): Promise<void> => {
+  try {
+    // Create the Express application
+
+    // Middleware for parsing JSON request body
+    app.use(bodyParser.json({ limit: '5mb' }));
+
+    app.use(bodyParser.urlencoded({ limit: '5mb', extended: true, parameterLimit: 50000 }));
```
Comment (HarishGangula) [inline src/app.ts:20]: urlencoded used multiple times with same settings

File: src/app.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    };
+
+    // Handle uncaught exceptions and unhandled rejections
+    const unexpectedErrorHandler = (error: Error): void => {
+      logger.error('error', error);
+
+      exitHandler();
+    };
+
+    process.on('uncaughtException', unexpectedErrorHandler);
```
Comment (HarishGangula) [inline src/app.ts:72]: `uncaughtException`, `unhandledRejection` should be outside of the method

File: src/config/config.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  log: {
+    day: get(envVars, 'LOG_FILE_DAY', '14d'),
+    isEnable: isEqual(get(envVars, 'LOG_FILE_ENABLE', 'true'), 'true'),
+    name: get(envVars, 'LOG_FILE_NAME', 'AML-log'),
+    size: get(envVars, 'LOG_FILE_SIZE', '20m'),
+    zippedArchive: isEqual(get(envVars, 'LOG_FILE_ZIP_ARCHIVE', 'false'), 'true'),
+  },
+  envPort: get(envVars, 'AML_SERVICE_APPLICATION_PORT', '4000'),
+  applicationEnv: get(envVars, 'AML_SERVICE_APPLICATION_ENV', 'development'),
+  appVersion: get(envVars, 'AML_SERVICE_APP_VERSION', '1.1'),
```
Comment (HarishGangula) [inline src/config/config.ts:17]: version should be 1.0

File: src/config/databaseConfig.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+} = appConfiguration;
+
+const AppDataSource = new Sequelize({
+  dialect: 'postgres',
+  host: host,
+  port: port,
+  username: user,
+  password: password,
+  database: name,
+  models: [__dirname + '/models/*.ts'],
```
Comment (HarishGangula) [inline src/config/databaseConfig.ts:16]: use path package here

**Review by HarishGangula — CHANGES_REQUESTED** [review]
1. Fix the spelling mistakes
2. Removed all the dependencies
3. Proper validation for the request body

File: package.json (hunk 1 above)
Comment (vishwab1) [reply package.json]: Removed from file

File: package.json (hunk 2 above)
Comment (vishwab1) [reply package.json]: removed multer and pg move to dev dependencies

File: src/config/databaseConfig.ts (hunk 3 above)
Comment (vishwab1) [reply src/config/databaseConfig.ts]: Removed

File: package.json (hunk 4 above)
Comment (vishwab1) [reply package.json]: Removed

File: src/app.ts (hunk 5 above)
Comment (vishwab1) [reply src/app.ts]: kept only express json and urlencoded

File: src/app.ts (hunk 6 above)
Comment (vishwab1) [reply src/app.ts:66]: kept outside method

File: src/config/databaseConfig.ts (hunk 8 above)
Comment (vishwab1) [reply src/config/databaseConfig.ts]: Used the Path packages

File: src/config/config.ts (hunk 7 above)
Comment (vishwab1) [reply src/config/config.ts]: changed to 1.0

File: src/models/tenant.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    },
+    identifier: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    name: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    type: {
```
Comment (HarishGangula) [inline src/models/tenant.ts:20]: add the allowed values in the tenant types, default of user not providing

File: src/utils/response.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  };
+};
+
+const generateErrorResponse = (id: string, err: string, errmsg: string, responseCode: string): ResponseObject => ({
+  id,
+  ver: appVersion,
+  ts: new Date().toISOString(),
+  params: {
+    resmsgid: uuid.v4(),
+    msgid: uuid.v4(),
```
Comment (HarishGangula) [inline src/utils/response.ts:48]: this user sent msgid

File: src/controllers/TenantCreate/TenantCreate.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  try {
+    //validating the schema
+    const isRequestValid: Record<string, any> = schemaValidation(requestBody, tenantCreateJson);
+    if (!isRequestValid.isValid) {
+      const code = 'TENANT_INVALID_INPUT';
+      logger.error({ code, apiId, msgid, requestBody, message: isRequestValid.message });
+      return res.status(httpStatus.BAD_REQUEST).json(errorResponse(apiId, msgid, httpStatus.BAD_REQUEST, isRequestValid.message, code));
+    }
+
+    if (dataBody.board_id && dataBody.board_id.length > 0) {
```
Comment (HarishGangula) [inline src/controllers/TenantCreate/TenantCreate.ts:27]: This check is not required

File: src/controllers/TenantCreate/TenantCreate.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+import * as uuid from 'uuid';
+import { getBoardsByIds } from '../../services/board';
+
+export const apiId = 'api.tenant.create';
+
+const tenantCreate = async (req: Request, res: Response) => {
+  const requestBody = _.get(req, 'body');
+  const msgid = _.get(req, ['body', 'params', 'msgid']);
+  const dataBody = _.get(req, 'body.request');
+  try {
```
Comment (HarishGangula) [inline src/controllers/TenantCreate/TenantCreate.ts:18]: Remove try catch

File: src/controllers/TenantCreate/TenantCreate.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        const code = 'BOARD_NOT_FOUND';
+        logger.error({ code, apiId, msgid, requestBody, message: 'board IDs do not exist' });
+        return res.status(httpStatus.BAD_REQUEST).json(errorResponse(apiId, msgid, httpStatus.BAD_REQUEST, 'Board IDs do not exist', code));
+      }
+    }
+
+    const tenantInsertData = {
+      ...dataBody,
+      identifier: uuid.v4(),
+      status: 'live',
```
Comment (HarishGangula) [inline src/controllers/TenantCreate/TenantCreate.ts:44]: move live to enum

File: src/controllers/TenantRead/GetSingleTenant.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+    //validating tenant is exist
+    if (_.isEmpty(tenantDetails)) {
+      const code = 'TENANT_NOT_EXISTS';
+      logger.error({ code, apiId, message: `Tenant not exists` });
+      return res.status(httpStatus.NOT_FOUND).json(errorResponse(apiId, msgid, httpStatus.NOT_FOUND, `Tenant does not exists `, code));
+    }
+
+    logger.info({ apiId, message: `Tenant Read Successfully` });
+    res.status(httpStatus.OK).json(successResponse(apiId, msgid, tenantDetails));
```
Comment (HarishGangula) [inline src/controllers/TenantRead/GetSingleTenant.ts:32]: Get the clarification on `id` and `identifier`

File: src/services/tenant.ts [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    await transact.commit();
+    return { error: false, updateTenant };
+  } catch (error: any) {
+    const errorMessage = error?.message || 'failed to update a record';
+    return { error: true, message: errorMessage };
+  }
+};
+
+//get Single tenant by id
+export const getTenantById = async (id: number): Promise<any> => {
```
Comment (HarishGangula) [inline src/services/tenant.ts:38]: take status as param and default it to live

File: src/services/tenant.ts [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  try {
+    const whereClause: Record<string, any> = { id };
+    whereClause.is_active = true;
```
Comment (HarishGangula) [inline src/services/tenant.ts:41]: move this line to above only

File: src/controllers/TenantRead/GetSingleTenant.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import { getTenant } from '../../services/tenant';
+
+export const apiId = 'api.tenant.read';
+
+const ReadSingleTenant = async (req: Request, res: Response) => {
+  const tenant_id = parseInt(_.get(req, 'params.tenant_id'));
+  const msgid = _.get(req, ['body', 'params', 'msgid']);
+
+  try {
+    const getTenantInfo = await getTenant(tenant_id);
```
Comment (HarishGangula) [inline src/controllers/TenantRead/GetSingleTenant.ts:15]: always get live tenant which is active

File: src/controllers/TenantSearch/searchTenantValidationSchema.json [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+      },
+      "required": ["msgid"],
+      "additionalProperties": false
+    },
+    "request": {
+      "type": "object",
+      "properties": {
+        "filters": {
+          "type": "object",
+          "description": "Object containing search criteria. Can include any fields defined in the Tenant model."
```
Comment (HarishGangula) [inline src/controllers/TenantSearch/searchTenantValidationSchema.json:29]: add specific filters

Comment (HarishGangula) [inline src/controllers/TenantSearch/searchTenantValidationSchema.json:29]: put specific filters

File: src/controllers/TenantUpdate/updateTenant.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    // Validate tenant existence
+    const isTenantExists = await checkTenantExists(tenant_id);
+
+    if (!isTenantExists) {
+      const code = 'TENANT_NOT_EXISTS';
+      logger.error({ code, apiId, requestBody, message: `Tenant not exists` });
+      return res.status(httpStatus.NOT_FOUND).json(errorResponse(apiId, msgid, httpStatus.NOT_FOUND, `Tenant not exists`, code));
+    }
+
+    if (dataBody.board_id && dataBody.board_id.length > 0) {
```
Comment (HarishGangula) [inline src/controllers/TenantUpdate/updateTenant.ts:36]: Not required

File: src/controllers/TenantUpdate/updateTenant.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+  } catch (error) {
+    const err = error instanceof Error;
+    const code = _.get(error, 'code', 'TENANT_UPDATE_FAILURE');
+    logger.error({ error, apiId, code, requestBody });
+    return res.status(httpStatus.INTERNAL_SERVER_ERROR).json(errorResponse(apiId, msgid, httpStatus.INTERNAL_SERVER_ERROR, err ? error.message : '', code));
+  }
+};
+
+// Helper functions
+const checkTenantExists = async (tenant_id: number): Promise<boolean> => {
```
Comment (HarishGangula) [inline src/controllers/TenantUpdate/updateTenant.ts:61]: Move this to service file

File: src/services/board.ts [hunk 21]
```diff
@@ -0,0 +1,13 @@
+import { boardMaster } from '../models/boardMaster';
+
+//get board by ids
+export const getBoardsByIds = async (board_ids: number[]): Promise<any> => {
```
Comment (HarishGangula) [inline src/services/board.ts:4]: rename this to getBoards

File: src/services/tenant.ts [hunk 22]
```diff
@@ -0,0 +1,76 @@
+import { Tenant } from '../models/tenant';
+import { AppDataSource } from '../config';
+import { Optional } from 'sequelize';
+import { UpdateTenant } from '../types/TenantModel';
+import _ from 'lodash';
+
+//create service for tenant
+export const createTenant = async (req: Optional<any, string> | undefined): Promise<any> => {
+  const transact = await AppDataSource.transaction();
```
Comment (HarishGangula) [inline src/services/tenant.ts:9]: Remove transaction from all methods

**Review by HarishGangula — CHANGES_REQUESTED** [review]

File: src/controllers/TenantCreate/TenantCreate.ts (hunk 12 above)
Comment (vishwab1) [reply src/controllers/TenantCreate/TenantCreate.ts]: Removed added the promise method

File: src/controllers/TenantCreate/TenantCreate.ts (hunk 11 above)
Comment (vishwab1) [reply src/controllers/TenantCreate/TenantCreate.ts]: removed this check

File: src/utils/response.ts (hunk 10 above)
Comment (vishwab1) [reply src/utils/response.ts]: changed msgid

File: src/controllers/TenantCreate/TenantCreate.ts (hunk 13 above)
Comment (vishwab1) [reply src/controllers/TenantCreate/TenantCreate.ts]: created the new folder enum and kept a file

File: src/controllers/TenantSearch/searchTenantValidationSchema.json (hunk 18 above)
Comment (vishwab1) [reply src/controllers/TenantSearch/searchTenantValidationSchema.json]: Added the name isactive and status now

File: src/services/board.ts (hunk 21 above)
Comment (vishwab1) [reply src/services/board.ts]: change to getBoards

File: src/services/tenant.ts (hunk 22 above)
Comment (vishwab1) [reply src/services/tenant.ts]: Removed

File: src/app.ts [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+    app.use(amlErrorHandler);
+
+    // 404 handler for unknown API requests
+    app.use((req: Request, res: Response, next: NextFunction) => {
+      next({ statusCode: NOT_FOUND, message: 'Not found' });
+    });
+
+    // Database connection
+    await AppDataSource.sync()
```
Comment (HarishGangula) [inline src/app.ts:59]: Remove the then and catch

**Review by HarishGangula — APPROVED** [review]

**Pull request merged**